### DISCUSSION TITLE
Page layout

### DISCUSSION THREAD

**Discussion by beres** [Hieromon/AutoConnect#497] [Ideas]
Hi @Hieromon,

Thank you for this library, I would like to change the layout of some pages.
Initially I wanted to change the colors of the _PAGE_CONFIG buttons which are not skinnable, by inserting a new label and sending a patch.

Then I thought it would be nice to be able to customize the layout of the pages, what could be the correct approach in your opinion?
You might be interested?

**Comment by Hieromon** [comment]
This topic moves to the Discussions. @beres It's not easy to add flexibility to the layout of AutoConnect's embedded pages, but there may be an opportunity to do so. Hope to continue in the Discussions.

**Reply by beres** [reply]
Hi @Hieromon ,

You're right, the fastest solution to achieve my goal was fork the library and customize AutoConnectPage, but this is far from to be the optimal or easy solution to make some little changes on layout and style.

I like this library it's well written and documented, but for my use case it has too many options, i don't use any AutoconnectElement  or custom settings. So for me probably replace AutoConnectPage whit some callback to read the page data would be enough.

Probably an option compatible with all the functionality of AutoCommect could be create some API to permit to use full customized layout, and find a way to deactivate AutoConnectPage to save some space.

B

**Comment by Hieromon** [comment]
@beres Sorry for the late reply. I have spent less time thinking about your suggestions because I have been prioritizing work on revising HTML generation for AutoConnectElement for the past few days.

The following is a simplified block diagram of the AutoConnect structure. It may help us in making your suggestion more concrete.

![AutoConnectDiagram](https://user-images.githubusercontent.com/12591771/176371106-56b6ed1b-6fcd-4ea3-9b67-28f0de9e0be9.svg)

According to PlatformIO's static diagnostics, the largest build size is the AutoConnectElements classes of the AutoConnectAux handler, which is an instantiation of the AutoConnectElements template. The next largest are the AutoConnect core and AutoConnectAux handler, followed by AutoConnectPage.cpp.

For use cases that do not require AutoConnectElements, it is indeed a wasteful structure; AutoConnectElemetJsonImpl is reduced by using the [**AUTOCONNECT_NOUSE_JSON**](https://hieromon.github.io/AutoConnect/faq.html#how-do-i-detach-the-ardunojson) macro definition, but still not a reduction. As you pointed out, AutoConnectPage.cpp is still the big one. Among them, the LUXBAR has a CSS size of 2.7KB.

It may be possible to completely isolate AutoConnectAux by templating the AutoConnect class and instantiating it so that it does not contain an API for custom web pages. 

I will look at a few more ideas on how to disconnect AutoConnectPage.

![inspect](https://user-images.githubusercontent.com/12591771/176387684-a287a838-c397-420e-8088-32cca9e44934.png)

![inspect_detail](https://user-images.githubusercontent.com/12591771/176387863-78173963-84c5-4dee-a266-eaec2b87625b.png)
